crates/stages/src/stages/index_storage_history.rs

```diff
@@ -80,12 +86,35 @@ impl<DB: Database> Stage<DB> for IndexStorageHistoryStage {
             return Ok(ExecOutput::done(input.checkpoint()))
         }
 
-        let (range, is_final_range) = input.next_block_range_with_threshold(self.commit_threshold);
-
-        let indices = provider.changed_storages_and_blocks_with_range(range.clone())?;
-        provider.insert_storage_history_index(indices)?;
-
-        Ok(ExecOutput { checkpoint: StageCheckpoint::new(*range.end()), done: is_final_range })
+        let range = input.next_block_range();
+
+        let collector = collect_history_indices::<
+            _,
+            reth_db::tables::StorageChangeSets,
+            reth_db::tables::StoragesHistory,
+            _,
+        >(
+            provider.tx_ref(),
+            BlockNumberAddress::range(range.clone()),
+            |AddressStorageKey((address, storage_key)), highest_block_number| {
+                StorageShardedKey::new(address, storage_key, highest_block_number)
+            },
+            |(key, value)| (key.block_number(), AddressStorageKey((key.address(), value.key))),
+        )?;
+
+        let append_only = provider.count_entries::<reth_db::tables::StoragesHistory>()?.is_zero();
```

Comment:
do we gain anything from the ETL here if `append_only == false`?

Reply:
yes, insert sorted vs insert unsorted

Reply:
https://github.com/paradigmxyz/reth/pull/1130#issuecomment-1418642755